### PR TITLE
agent: fix device ID conflicts

### PR DESCRIPTION
**Pull request by bergwolf**
As stated in https://github.com/kata-containers/agent/issues/834 and https://github.com/kata-containers/agent/issues/835, there are two cases that the device IDs can conflict and the rust agent does not handle them correctly.

1. Two devices have the same device IDs on the host
2. Device A's device ID is identical to device B's device ID on the host

The PR handles both of them and adds UTs to make sure it works as expected.

/cc @dgibson

### CODE REVIEW

**Comment by dgibson** [comment]
Hm, so.  Your patches fix the same problems in a more minimal way.  That's nicer in some ways (no extra map data structure), less nice in others (it puts extra information in the long-lived oci spec structure that's only relevant for a shortish window).

The larger issue for me, is that while your approach solves these specific conflicts, it doesn't fix the other problem I needed to solve for my SR-IOV work: I need to allow a device to have a different *name* in the inner container from the outer oci spec as well as different major:minor numbers.

Both your approach and mine are a bit ugly in that there's a non-trivial window where the spec structure is in a weird "half translated" state where you need to look at extra information to interpret which pieces are in host format and which are in guest format.  Long term, I think a better approach would be to completely remove the devices sections of the spec on the runtime side, put all the information we need into the sideband devices information, then rebuild the spec devices section in the agent from that.  That way we can be more confident that we don't have unnecessary host-specific information leaking into the VM.

But, that's going to require some substantial reworking, so we need to figure our what to do in the meantime.  Let me see what I can come up with today.

**Comment by bergwolf** [comment]
> The larger issue for me, is that while your approach solves these specific conflicts, it doesn't fix the other problem I needed to solve for my SR-IOV work: I need to allow a device to have a different name in the inner container from the outer oci spec as well as different major:minor numbers.

IIUC your use case would require modification to how the current mapping between host devices and guest devices are done. And it means that we need to change the agent protocol and thus not a suitable change for 2.0 (for which we have decided not to change the agent protocol). We can support it in future releases but not the current one.

>  Long term, I think a better approach would be to completely remove the devices sections of the spec on the runtime side, put all the information we need into the sideband devices information, then rebuild the spec devices section in the agent from that. That way we can be more confident that we don't have unnecessary host-specific information leaking into the VM.

Yes, I agree. But again at the current stage, we do not want to change the agent protocol for the 2.0 release.

**Comment by dgibson** [comment]
> > The larger issue for me, is that while your approach solves these specific conflicts, it doesn't fix the other problem I needed to solve for my SR-IOV work: I need to allow a device to have a different name in the inner container from the outer oci spec as well as different major:minor numbers.
> 
> IIUC your use case would require modification to how the current mapping between host devices and guest devices are done. And it means that we need to change the agent protocol and thus not a suitable change for 2.0 (for which we have decided not to change the agent protocol). We can support it in future releases but not the current one.

Actually, it doesn't require a protocol change.  The fix for this bug, as I implemented it for Kata 1 also allows what I need.  We still match the grpc devices to the OCI spec devices using the "outer" container path, but once we've established where they live in the "inner" OCI spec, we can then change that path for the inner container.

> > Long term, I think a better approach would be to completely remove the devices sections of the spec on the runtime side, put all the information we need into the sideband devices information, then rebuild the spec devices section in the agent from that. That way we can be more confident that we don't have unnecessary host-specific information leaking into the VM.
> 
> Yes, I agree. But again at the current stage, we do not want to change the agent protocol for the 2.0 release.

Fair enough, but as noted we don't need one for this.

Although your fix here is a little more elegant than mine, I'd actually prefer to keep the same structure as the Kata 1 / Go fix for now, since it will make forward ports that depend on it (I have several in my queue) easier.  I've filed #931 for that approach.

File: src/agent/src/device.rs [hunk 1]
```diff
@@ -218,46 +218,73 @@ fn update_spec_device_list(device: &Device, spec: &mut Spec) -> Result<()> {
...
+// 1. find out linux device matching dev_container_path and update its (major, minor) to
+//    match given (guest_major, guest_minor)
+// 2. look for matching device cgroups constraints in linux.resources and update their (major,
+//    minor) to match giving (guest_major, guest_minor)
+fn update_one_spec_device(
+    linux: &mut Linux,
+    dev_container_path: &str,
+    dev_type: &str,
+    guest_major: i64,
+    guest_minor: i64,
```
Comment (jodh-intel) [inline src/agent/src/device.rs:250]: OOI, why not `u64`?

**Review by jodh-intel — APPROVED** [review]
Thanks @bergwolf.

lgtm

**Comment by bergwolf** [comment]
replaced by #931

**Pull request closed without merging**